src/main/java/net/dv8tion/jda/api/utils/Once.java

```diff
@@ -0,0 +1,214 @@
+package net.dv8tion.jda.api.utils;
+
+import net.dv8tion.jda.api.JDA;
+import net.dv8tion.jda.api.events.GenericEvent;
+import net.dv8tion.jda.api.hooks.EventListener;
+import net.dv8tion.jda.api.hooks.SubscribeEvent;
+import net.dv8tion.jda.api.utils.concurrent.Task;
+import net.dv8tion.jda.internal.utils.Checks;
+import net.dv8tion.jda.internal.utils.concurrent.task.GatewayTask;
+
+import javax.annotation.CheckReturnValue;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+/**
+ * Helper class to listen to an event, once.
+ *
+ * @param <E> Type of the event listened to
+ *
+ * @see   JDA#listenOnce(Class)
+ */
+public class Once<E extends GenericEvent> implements EventListener
+{
+    private final JDA jda;
+    private final Class<E> eventType;
+    private final List<Predicate<? super E>> filters;
+    private final CompletableFuture<E> future;
+    private final GatewayTask<E> task;
+    private final ScheduledFuture<?> timeoutFuture;
+    private final Runnable timeoutCallback;
+
+    private Once(@Nonnull Once.Builder<E> builder)
+    {
+        this.jda = builder.jda;
+        this.eventType = builder.eventType;
+        this.filters = new ArrayList<>(builder.filters);
+        this.timeoutCallback = builder.timeoutCallback;
+
+        this.future = new CompletableFuture<>();
+        this.task = createTask(future);
+        this.timeoutFuture = scheduleTimeout(builder.timeout, future);
+    }
+
+    @Nonnull
+    private GatewayTask<E> createTask(@Nonnull CompletableFuture<E> future)
+    {
+        final GatewayTask<E> task = new GatewayTask<>(future, () ->
+        {
+            // On cancellation, throw cancellation exception and cancel timeout
+            jda.removeEventListener(this);
+            future.completeExceptionally(new CancellationException());
+            if (timeoutFuture != null)
+                timeoutFuture.cancel(false);
+        });
+        task.onSetTimeout(e ->
+        {
+            throw new UnsupportedOperationException("You must set the timeout on Once.Builder#timeout");
+        });
+        return task;
+    }
+
+    @Nullable
+    private ScheduledFuture<?> scheduleTimeout(@Nullable Duration timeout, @Nonnull CompletableFuture<E> future)
+    {
+        if (timeout == null) return null;
+
+        return jda.getGatewayPool().schedule(() ->
+        {
+            // On timeout, throw timeout exception and run timeout callback
+            jda.removeEventListener(this);
+            future.completeExceptionally(new TimeoutException());
+            if (timeoutCallback != null)
```

Comment:
Use the return value of `completeExceptionally` to check if the future was actually completed by this call. This way you prevent race-conditions where both callbacks can run. Do the same below for the success cases.